More edge cases for FBBT

michaelbynum opened the pull request:
## Changes proposed in this PR:
This PR handles some more edge cases in the interval arithmetic in the FBBT module. In particular, 

```
div(0, -1e-16, 0, 0, 1e-8)
```

currently raises an exception. In reality, this should return `(-inf, inf)`.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

jsiirola commented:
I am generally concerned by the number of inconsistencies between `div(xl, xu, yl, yu)` and `mul(xl, xu, *inv(yl, yu))` that I don't understand.  For example, given `tol = 1e-10`:
`xl`, `xu` | `yl`,  `yu` | `div(xl, xu, yl, yu, tol)` | `mul(xl, xu, *inv(yl, yu, tol))`
------|------|------------------|-----------------------
0, 0 | 0, 1e-11 | -inf, inf | 0, 0
0, 0 | 0, -1e-12 | -inf, inf | IntervalException
0, 0 | 0, -1e-10 | -inf, inf | InfeasibleConstraintException
0, 5e-11 | 0, 1e-11 | -inf, inf | 0, inf
0, -1e-12 | 0, 1e-12 | -inf, inf | -inf, 0
1e-12, 0 | -1e-12, -1e-12 | -inf, inf | -inf, -0
1e-12, 1e-12 | 0, 5e-11 | -inf, inf | 0.02, inf
1e-12, 1e-12 | 1e-12, 1e-12 | -inf, inf | 1, inf
1e-12, 1e-12 | -1e-12, -1e-12 | -inf, inf | -inf, -1
5e-11, 5e-11 | -1e-12, -1e-12 | -inf, inf | -inf, -50

If you run the following "exhaustive" comparison:
```python
import sys
import itertools
from pyomo.contrib.fbbt.interval import div, mul, inv

tol = 1e-10
numbers = (0, 1e-12, 5e-11, 1e-10, 1e-11, -1e-12, -5e-11, -1e-10, -1e-11)

ok = 0
fail = 0
for xl, xu, yl, yu in itertools.product(numbers, numbers, numbers, numbers):
    try:
        d = "%g, %g" % div(xl, xu, yl, yu, tol)
    except:
        d = sys.exc_info()[0].__name__
    try:
        mi = "%g, %g" % mul(xl, xu, *inv(yl, yu, tol))
    except:
        mi = sys.exc_info()[0].__name__
    if d == mi:
        ok += 1
        continue
    fail += 1
    print("%g, %g | %g, %g | %s | %s" % (
        xl, xu, yl, yu, d, mi))

print(f"ok={ok}, fail={fail}")
```
you get
```
ok=2756, fail=3805
```

michaelbynum commented:
@jsiirola Thanks for this. It was really helpful. Most of the differences were undesirable, but reasonable. There was one particularly bad bug that this highlighted, which I have fixed and added a test for. I was also able to make some minor modifications so that `div(xl, xu, yl, yu, tol)` just calls `mul(xl, xu, *inv(yl, yu, tol))`. The primary change I had to make to enable this was to ensure `inv` returns `(-inf, inf)` instead of raising an exception, which is very reasonable, I think.

jsiirola commented on pyomo/contrib/fbbt/interval.py:
```diff
@@ -74,11 +80,7 @@ def inv(xl, xu, feasibility_tol):
-        ub = inf
-    else:
-        lb, ub = mul(xl, xu, *inv(yl, yu, feasibility_tol))
+    lb, ub = mul(xl, xu, *inv(yl, yu, feasibility_tol))
     return lb, ub
```
It is slightly faster to just `return mul(...` (Python doesn't have to unpack the tuple, update the `locals()`, recreate the tuple, and increment / decrement references)
